Add sgx-show

npmccallum opened the pull request:
This program exists to gather and display data relating to SGX hardware, SGX BIOS enablement, SGX kernel drivers and potentially SGX-related software for use in debugging.

steveeJ commented on sgx-show/src/main.rs:
```diff
@@ -0,0 +1,176 @@
+//!
+//! It works by running a hierarchical tree of tests and displaying the
+//! information gathered therefrom. If a parent test fails to validate,
+//! the child test are still called to produce their output, but do not
```
```suggestion
//! the child tests are still called to produce their output, but do not
```

npmccallum replied:
Fixed.

steveeJ commented on sgx-show/src/main.rs:
```diff
@@ -0,0 +1,176 @@
+use data::cpuid::{CpuId, CpuInfo, ExtFeat, SgxAttr, SgxCaps};
+use sink::{bit::Bit, debug::Debug, mask::Mask};
+
+pub trait Convert<T>: Sized {
```
Could we use the existing https://doc.rust-lang.org/std/convert/trait.TryInto.html instead here?

npmccallum replied:
Originally, no; which is why I created `Convert`. But because of other changes that is now possible. I have moved this to `TryFrom`. So: fixed.

steveeJ commented on sgx-show/src/data/cpuid.rs:
```diff
@@ -0,0 +1,81 @@
+
+impl<T: Leaf> CpuId<T> {
+    pub fn eax(&self) -> u32 {
+        (self.0 >> 96) as u32
```
```suggestion
        (self.0 >> 3 * 32) as u32
```

npmccallum replied:
Fixed.

steveeJ commented on sgx-show/src/data/cpuid.rs:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    pub fn ebx(&self) -> u32 {
+        (self.0 >> 64) as u32
```
```suggestion
        (self.0 >> 2 * 32) as u32
```

npmccallum replied:
Fixed.

steveeJ commented on sgx-show/src/data/cpuid.rs:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    pub fn ecx(&self) -> u32 {
+        (self.0 >> 32) as u32
```
```suggestion
        (self.0 >> 1 * 32) as u32
```

npmccallum replied:
Fixed.

steveeJ commented on sgx-show/src/data/cpuid.rs:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    pub fn edx(&self) -> u32 {
+        (self.0 >> 00) as u32
```
```suggestion
        (self.0 >> 0 * 32) as u32
```

npmccallum replied:
Fixed.

npmccallum commented:
@steveeJ I've addressed all your issues. Is it possible for you to remove the "changes requested" state without approving the whole PR? (Honest question.)

lkatalin commented on sgx-show/src/sink/bit.rs:
```diff
@@ -0,0 +1,14 @@
+    type Type = CpuId<T>;
+
+    fn test(&self, data: &Self::Type) -> bool {
+        (self.reg)(data) & (1 << self.bit) != 0
```
I don't totally understand this line and `(self.reg)(data)`. `data` is a reference to a `CpuId<T>` and `reg` is a function to turn that into a `u32`? What do the parens do here?

npmccallum replied:
`self.reg` is a function pointer. So we are calling the function pointed to by `self.reg`. This is a function to extract one particular register from the CpuId struct. See, for example: https://github.com/enarx/enarx/pull/97/files#diff-187ebd96da7420d347a4eae0e18468b6R124

lkatalin commented on sgx-show/src/main.rs:
```diff
@@ -0,0 +1,174 @@
+use sink::{bit::Bit, debug::Debug, mask::Mask};
+
+impl TryFrom<CpuId<SgxCaps>> for misc::MiscSelect {
+    type Error = ();
```
Why is the Error type `()` here? Shouldn't there be more information returned for error handling?

npmccallum replied:
The error is always discarded: https://github.com/enarx/enarx/pull/97/files#diff-8c921bd2af53e89a5b8942f9d0f785d7R44

lkatalin replied:
Why do we want to discard the error by converting the `Result` into an `Option` there?

npmccallum replied:
Because we don't report specific errors at this time. Basically, the test succeeds only if all of the conditions apply:
1. Data collection succeeds.
2. Data type conversion succeeds.
3. Data evaluation succeeds.

If all three conditions are true, we get a green check. Otherwise, we get a red 'x'.

lkatalin commented on sgx-show/src/main.rs:
```diff
@@ -0,0 +1,174 @@
+        name: "Intel CPU",
+        data: Box::new(CpuInfo),
+        sink: Box::new(Mask {
+            mask: 0x00000000ffffffffffffffffffffffffu128,
```
I assume these values come from Intel documentation. Is the value masked because only the first bits are relevant? Or is the mask for some other purpose? Will it potentially change?

npmccallum replied:
See here for details: https://en.wikipedia.org/wiki/CPUID#EAX=0:_Highest_Function_Parameter_and_Manufacturer_ID

EAX contains the highest CPUID leaf. EBX, ECX and EDX contain the manufacturer identifier. We mask out EAX.

lkatalin commented on sgx-show/src/sink/mod.rs:
```diff
@@ -0,0 +1,15 @@
+    type Type;
+
+    fn info(&self, _value: &Self::Type) -> Option<String> {
+        None
```
It seems like this will *always* return `None`; is there something I'm missing about how this is used?

npmccallum replied:
This is a default function which will be overridden by specific implementations of `Sink`. The default outputs nothing.

lkatalin replied:
Ahh, I see. Ex. here https://github.com/enarx/enarx/pull/97/files#diff-28457ac1732bda7b910ee565ab93f102R12

npmccallum replied:
Exactly.
